### PR TITLE
feat(@rifat/miio): new type definition

### PR DESCRIPTION
**Pull request by Shaquu**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [x] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [x] Delete the package's directory.
- [x] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## rifat__miio ([<kbd>unpkg</kbd>](https://unpkg.com/browse/rifat__miio@latest/))
was missing the following properties:
1. models
1. infoFromHostname
1. browse
1. devices
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 9a11f7d666622b50e2bb8ad2c6b59413aaa485d2
</p>

File: types/rifat__miio/rifat__miio-tests.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { device } from '@rifat/miio';
+
+device({ address: '192.168.0.1' }); // $ExpectError
```
Comment (domdomegg) [inline types/rifat__miio/rifat__miio-tests.ts:3]: The docs suggest this should be valid if the token can be determined automatically or from storage

File: types/rifat__miio/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    changeFavoriteLevel(value: string): Promise<string>;
+
+    changeFanSpeed(value: string): Promise<string>;
+}
+
+export type MiioDevice = Device | AirPurifier3;
+
+export interface MiioDeviceOptions {
+    address: string;
+    token: string | Buffer;
```
Comment (domdomegg) [inline types/rifat__miio/index.d.ts:44]:
```suggestion
    token?: string | Buffer;
```

File: types/rifat__miio/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+// Definitions by: Tadeusz Wyrzykowski <https://github.com/Shaquu>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference types="node" />
+
+export interface Device {
+    on(eventName: string | symbol, listener: (...args: any[]) => void): this;
+}
+
+export interface AirPurifier3 extends Device {
```
Comment (domdomegg) [inline types/rifat__miio/index.d.ts:12]: Add overloaded .matches method here:

matches(type: "type:air-purifier"): true
matches(type: string): boolean

**Comment by Shaquu** [comment]
@domdomegg thanks for comments. Let's wait with that PR until #57897 is done.

**Review by domdomegg — CHANGES_REQUESTED** [review]
Thanks for updating to the latest master. Review remains as per previous comments, but looking good and only need a few tweaks to get this in :)

**Comment by Shaquu** [comment]
@domdomegg thanks for comments :)
I will rework it as I got my other types merged which might be useful here. 

So prepare for more work! :)

File: types/rifat__miio/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
+// Type definitions for @rifat/miio 2.0
+// Project: https://github.com/torifat/miio
+// Definitions by: Tadeusz Wyrzykowski <https://github.com/Shaquu>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference types="node" />
+
+export interface Device {
```
[the remark this answers is not in the document]
Comment (Shaquu) [reply types/rifat__miio/index.d.ts:8]: Implemented in abstract-things
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/8c1d141f157141b359ae29644e11fdafc5eac0ee/types/abstract-things/thing.d.ts#L28

File: types/rifat__miio/index.d.ts (hunk 3 above)
Comment (Shaquu) [reply types/rifat__miio/index.d.ts:12]: Will be added

File: types/rifat__miio/index.d.ts (hunk 2 above)
Comment (Shaquu) [reply types/rifat__miio/index.d.ts:44]: Will be added

File: types/rifat__miio/rifat__miio-tests.ts (hunk 1 above)
Comment (Shaquu) [reply types/rifat__miio/rifat__miio-tests.ts:3]: Will be resolved

File: types/rifat__miio/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+export interface MiioDeviceOptions {
+    address: string;
+    token?: string | Buffer;
+    port?: string;
+    model?: string;
+    withPlaceholder?: boolean;
+}
+
+declare module '@rifat/miio' {
+    function device<T extends MiioDevice>(options: MiioDeviceOptions): Promise<T>;
```
Comment (Shaquu) [inline types/rifat__miio/index.d.ts:51]: There is still an issue here with T being bad during tests.
Looking for solutions.

**Comment by Shaquu** [comment]
I am in isolation currently, I need more time to address the issues.

**Comment by Shaquu** [comment]
> @Shaquu I haven't seen any activity on this PR in more than three weeks, and it still has problems that prevent it from being merged. The PR will be closed on Mar 1st (in a week) if the issues aren't addressed.

Got it

**Comment by Shaquu** [comment]
@domdomegg I did few improvements. Please take a look :)

**Review by domdomegg — APPROVED** [review]

File: types/rifat__miio/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    address: string;
+    token?: string | Buffer;
+    port?: string;
+    model?: string;
+    withPlaceholder?: boolean;
+}
+
+declare module '@rifat/miio' {
+    function device(options: MiioDeviceOptions): Promise<MiioDevice>;
+}
```
Comment (peterblazejewicz) [inline types/rifat__miio/index.d.ts:52]:
```suggestion
export function device(options: MiioDeviceOptions): Promise<MiioDevice>;
```
just for consistency with other exports (direct, on module level)

**Review by peterblazejewicz — APPROVED** [review]

**Comment by Shaquu** [comment]
Ready to merge

**Pull request merged**